### PR TITLE
chore: Added support for parallel run in nightwatch cucumber

### PR DESCRIPTION
**Pull request by sauravdas1997**

### CODE REVIEW

File: src/utils/helper.js [hunk 1]
```diff
@@ -471,3 +485,43 @@ exports.uploadPending = async (
 exports.shutDownRequestHandler = async () => {
   await requestQueueHandler.shutdown();
 };
+
+exports.getUniqueIdentifierForCucumber = (pickle) => {
```
Comment (sriteja777) [inline src/utils/helper.js:489]: Are we using this function?

**Review by sriteja777 — CHANGES_REQUESTED** [review]

File: nightwatch/globals.js [hunk 2]
```diff
@@ -1,15 +1,89 @@
...
 const testObservability = new TestObservability();
 
 const nightwatchRerun = process.env.NIGHTWATCH_RERUN_FAILED;
 const nightwatchRerunFile = process.env.NIGHTWATCH_RERUN_REPORT_FILE;
+const _tests = {};
+const _testCasesData = {};
+
+const registerListeners = () => {
+  process.removeAllListeners(`bs:addLog:${process.pid}`);
+};
```
[the remark this answers is not in the document]
Comment (sauravdas1997) [reply nightwatch/globals.js:20]: done

File: src/utils/constants.js [hunk 3]
```diff
@@ -7,3 +7,11 @@ exports.RERUN_FILE = 'rerun.json';
 exports.DEFAULT_WAIT_TIMEOUT_FOR_PENDING_UPLOADS = 5000;
 exports.DEFAULT_WAIT_INTERVAL_FOR_PENDING_UPLOADS = 100;
 exports.CUSTOM_REPORTER_CALLBACK_TIMEOUT = 3600000;
+exports.consoleHolder = Object.assign({}, console);
+exports.PID_MAPPING_REGEX = /^TEST-OBSERVABILITY-PID-TESTCASE-MAPPING-[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/;
```
[the remark this answers is not in the document]
Comment (sauravdas1997) [reply src/utils/constants.js:11]: done

File: src/utils/helper.js (hunk 1 above)
Comment (sauravdas1997) [reply src/utils/helper.js:489]: removed

File: src/utils/logPatcher.js [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+class LogPatcher extends Transport {
+  constructor(opts) {
+    super(opts);
+
+    ipc.config.id = IPC_SERVER_NAME;
+    ipc.config.retry = 1500;
+    ipc.config.silent = true;
+  
+    ipc.connectTo(IPC_SERVER_NAME, () => {
```
[the remark this answers is not in the document]
Comment (sauravdas1997) [reply src/utils/logPatcher.js:22]: added

File: src/testObservability.js [hunk 5]
```diff
@@ -367,6 +367,109 @@ class TestObservability {
...
+        steps: steps,
+        examples: examples
+      }
+    };
+
+    try {
+      if (eventType === 'TestRunFinished') {
+        const currentSessionCapabilities = reportData.session[args.envelope.testCaseStartedId];
+        const sessionCapabilities = currentSessionCapabilities.capabilities;
+        if ((sessionCapabilities) && (args.envelope.testCaseStartedId === currentSessionCapabilities.testCaseStartedId)) {
```
[the remark this answers is not in the document]
Comment (sauravdas1997) [reply src/testObservability.js:408]: added

File: nightwatch/globals.js [hunk 6]
```diff
@@ -45,10 +119,189 @@ module.exports = {
...
+          ...args.envelope
+        };
+        const reportData = args.report;
+        const testCaseId = reportData.testCaseStarted[args.envelope.id].testCaseId;
+        const pickleId = reportData.testCases.find((testCase) => testCase.id === testCaseId).pickleId;
+        const pickleData = reportData.pickle.find((pickle) => pickle.id === pickleId);
+        const gherkinDocument = reportData?.gherkinDocument.find((document) => document.uri === pickleData.uri);
+        const featureData = gherkinDocument.feature;
+        const uniqueId = uuidv4();
+        _tests['uniqueId'] = testCaseId;
```
[the remark this answers is not in the document]
Comment (sauravdas1997) [reply nightwatch/globals.js:155]: done

File: nightwatch/globals.js [hunk 7]
```diff
@@ -45,10 +119,189 @@ module.exports = {
...
+
+      const data  = {args: args, pid: process.pid};
+
+      ipc.config.id = IPC_SERVER_NAME;
+      ipc.config.retry = 1500;
+      ipc.config.silent = true;
+
+      ipc.connectTo(IPC_SERVER_NAME, async () => {
+        await ipc.of.browserstackTestObservability.emit(IPC_EVENTS.SCREENSHOT, data);
+      });
```
[the remark this answers is not in the document]
Comment (sauravdas1997) [reply nightwatch/globals.js:303]: added

File: nightwatch/globals.js [hunk 8]
```diff
@@ -73,6 +326,8 @@ module.exports = {
     try {
       testObservability.configure(settings);
       if (helper.isTestObservabilitySession()) {
+        settings.test_runner.options['require'] = 'node_modules/@nightwatch/browserstack/nightwatch/observabilityLogPatcherHook.js';
```
[the remark this answers is not in the document]
Comment (sauravdas1997) [reply nightwatch/globals.js:329]: done

File: nightwatch/globals.js [hunk 9]
```diff
@@ -45,10 +119,189 @@ module.exports = {
...
+      }
+    });
+
+    eventBroadcaster.on('TestCaseStarted', async (args) => {
+      if (!helper.isTestObservabilitySession()) {
+        return;
+      }
+      try {
+        // starting IPC 
+        startIPCServer();
```
[the remark this answers is not in the document]
Comment (sauravdas1997) [reply nightwatch/globals.js:144]: updated

File: package.json [hunk 10]
```diff
@@ -44,7 +44,10 @@
     "gitconfiglocal": "^2.1.0",
     "request": "^2.88.2",
     "strip-ansi": "^6.0.1",
-    "uuid": "^9.0.0"
+    "winston-transport": "^4.5.0",
+    "uuid": "^9.0.0",
+    "node-ipc": "9.1.1",
+    "@cucumber/cucumber": "^8.2.1"
```
[the remark this answers is not in the document]
Comment (sauravdas1997) [reply package.json:50]: done

File: nightwatch/observabilityLogPatcherHook.js [hunk 11]
```diff
@@ -0,0 +1,5 @@
+const {Before} = require('@cucumber/cucumber');
```
[the remark this answers is not in the document]
Comment (sauravdas1997) [reply nightwatch/observabilityLogPatcherHook.js:1]: done

File: nightwatch/globals.js [hunk 12]
```diff
@@ -45,10 +119,189 @@ module.exports = {
...
-    }
+  registerEventHandlers(eventBroadcaster) {
+
+    eventBroadcaster.on('TestFinished', (args) => {
+      if (!helper.isTestObservabilitySession()) {
+        return;
+      }
+      try {
+        if (typeof browser !== 'undefined') {
+          browser.execute(`browserstack_executor: {"action": "annotate", "arguments": {"type":"Annotation","data":"ObservabilitySync:${Date.now()}","level": "debug"}}`);
```
[the remark this answers is not in the document]
Comment (sauravdas1997) [reply nightwatch/globals.js:130]: Currently no way to map the hostname of session to a certain test id. Nightwatch isn't sending `testCaseStartedId` in `args.envelope`. Any exception occurred will get handled here.

File: src/testObservability.js [hunk 13]
```diff
@@ -409,6 +409,8 @@ class TestObservability {
           testData.integrations = {};
           const provider = helper.getCloudProvider(currentSessionCapabilities.host);
           testData.integrations[provider] = helper.getIntegrationsObject(sessionCapabilities, currentSessionCapabilities.sessionId);
+        } else {
+          Logger.error('Failed to upload integrations data');
```
Comment (sriteja777) [inline src/testObservability.js:413]: This log will get shown to user right? Can we instead add a debug log that logs only when debug mode is enabled.

File: nightwatch/globals.js [hunk 14]
```diff
@@ -1,15 +1,86 @@
...
+};
+
+const getUUIDFromPID = (pid) => {
+  if (_testCasesData[pid] === undefined) {return}
+  const testCaseStartedId = _testCasesData[pid];
+  const testCaseId = _testCasesData[testCaseStartedId].testCaseId;
+  if (_tests[testCaseId] === undefined) {return}
+
+  return _tests[testCaseId].uuid;
+};
```
Comment (harshit-bs) [inline nightwatch/globals.js:42]: The function looks good overall, but we could take advantage of optional chaining to make it more concise and elegant.

```suggestion
const getUUIDFromPID = (pid) => {
  const testCaseStartedId = _testCasesData[pid];
  const testCaseId = _testCasesData[testCaseStartedId]?.testCaseId;
  return _tests[testCaseId]?.uuid;
};
```

File: src/utils/logPatcher.js [hunk 15]
```diff
@@ -0,0 +1,87 @@
...
+};
+
+class LogPatcher extends Transport {
+  constructor(opts) {
+    super(opts);
+
+    try {
+      ipc.config.id = IPC_SERVER_NAME;
+      ipc.config.retry = 1500;
+      ipc.config.silent = true;
```
Comment (harshit-bs) [inline src/utils/logPatcher.js:21]: If possible, can we set default IPC config in a single location, enabling us to utilize it across multiple instances.

File: src/utils/helper.js [hunk 16]
```diff
@@ -471,3 +485,43 @@ exports.uploadPending = async (
 exports.shutDownRequestHandler = async () => {
   await requestQueueHandler.shutdown();
 };
+
+exports.getScenarioExamples = (gherkinDocument, scenario) => {
+  if ((scenario.astNodeIds && scenario.astNodeIds.length <= 1) || scenario.astNodeIds === undefined) {
+    return;
+  }
```
Comment (harshit-bs) [inline src/utils/helper.js:492]:
```suggestion
  if (!(scenario.astNodeIds?.length > 1)) {
    return;
  }
```

File: nightwatch/globals.js [hunk 17]
```diff
@@ -414,3 +390,25 @@ module.exports = {
...
+    startWorker(...args) {
+      const workerData  = super.startWorker(...args);
+      workerList = this.workers;
+      
+      return workerData;
+    }
+    
+    parseWorkerMessage(...args) {
+      if ([EVENTS.LOG, EVENTS.LOG_INIT].includes(args[1].eventType)) {return}
+      super.parseWorkerMessage(...args);
```
Comment (07souravkunda) [inline nightwatch/globals.js:409]: return  super.parseWorkerMessage(...args);

File: nightwatch/globals.js [hunk 18]
```diff
@@ -414,3 +390,25 @@ module.exports = {
...
+const cucumberPatcher = () => {
+  const Coordinator = helper.requireModule('@cucumber/cucumber/lib/runtime/parallel/coordinator.js');
+  class CoordinatorPatcher extends Coordinator.default {
+    constructor(...args) {
+      super(...args);
+    }
+    
+    startWorker(...args) {
+      const workerData  = super.startWorker(...args);
+      workerList = this.workers;
```
Comment (07souravkunda) [inline nightwatch/globals.js:402]: startworker runs once?

File: nightwatch/globals.js [hunk 19]
```diff
@@ -316,12 +292,12 @@ module.exports = {
       testObservability.configure(settings);
       if (helper.isTestObservabilitySession()) {
```
Comment (07souravkunda) [inline nightwatch/globals.js:292]: please add try..catch for observability stuff.

File: src/utils/logPatcher.js [hunk 20]
```diff
@@ -1,7 +1,15 @@
 const Transport = require('winston-transport');
-const {consoleHolder, PID_MAPPING_REGEX, IPC_SERVER_NAME, IPC_EVENTS} = require('./constants');
+const {consoleHolder, PID_MAPPING_REGEX, IPC_SERVER_NAME, EVENTS} = require('./constants');
 const ipc = require('node-ipc');
 const CrashReporter = require('./crashReporter');
+const TestObservability = require('../testObservability');
+const testObservability = new TestObservability();
+const helper = require('./helper');
+const eve = require('events');
```
Comment (07souravkunda) [inline src/utils/logPatcher.js:7]: move package imports to top

File: src/utils/logPatcher.js [hunk 21]
```diff
@@ -1,7 +1,15 @@
 const Transport = require('winston-transport');
-const {consoleHolder, PID_MAPPING_REGEX, IPC_SERVER_NAME, IPC_EVENTS} = require('./constants');
+const {consoleHolder, PID_MAPPING_REGEX, IPC_SERVER_NAME, EVENTS} = require('./constants');
 const ipc = require('node-ipc');
 const CrashReporter = require('./crashReporter');
+const TestObservability = require('../testObservability');
+const testObservability = new TestObservability();
```
Comment (07souravkunda) [inline src/utils/logPatcher.js:5]: move var declarations after imports.

File: nightwatch/globals.js [hunk 22]
```diff
@@ -16,73 +17,31 @@ const nightwatchRerun = process.env.NIGHTWATCH_RERUN_FAILED;
...
-const uploadTestLog = async (data) => {
-  try {
-    const {loggingData, pid} = data;
-    const uuid = getUUIDFromPID(pid);
-    testObservability.appendTestItemLog(loggingData, uuid);
-  } catch (error) {
-    CrashReporter.uploadCrashReport(error.message, error.stack);
+let currentTestUUID = '';
+let workerList = {};
+const {consoleHolder} = require('../src/utils/constants');
```
Comment (07souravkunda) [inline nightwatch/globals.js:21]: move file imports to top

**Review by 07souravkunda — CHANGES_REQUESTED** [review]

File: nightwatch/globals.js (hunk 17 above)
Comment (07souravkunda) [reply nightwatch/globals.js:409]: args[1]?.eventType

**Comment by 07souravkunda** [comment]
Fix unit tests

File: nightwatch/globals.js (hunk 18 above)
Comment (sauravdas1997) [reply nightwatch/globals.js:402]: no loss in workers

File: nightwatch/globals.js (hunk 22 above)
Comment (sauravdas1997) [reply nightwatch/globals.js:21]: done

File: nightwatch/globals.js (hunk 19 above)
Comment (sauravdas1997) [reply nightwatch/globals.js:292]: done

File: nightwatch/globals.js [hunk 23]
```diff
@@ -414,3 +390,25 @@ module.exports = {
 
   }
 };
+
+const cucumberPatcher = () => {
+  const Coordinator = helper.requireModule('@cucumber/cucumber/lib/runtime/parallel/coordinator.js');
```
[the remark this answers is not in the document]
Comment (sauravdas1997) [reply nightwatch/globals.js:394]: this method is called only for cucumber, so for safety wrapping a trycatch around it

**Review by 07souravkunda — APPROVED** [review]

**Review by gravityvi — APPROVED** [review]

**Pull request merged**